src/usr/share/gpgmailer/gpgmailer.py

```diff
@@ -15,156 +13,164 @@
-        fingerprint_list = []
-        for recipient in self.config['recipients']:
-            fingerprint_list.append(recipient['fingerprint'])
-        # Encrypt the message
-        encrypted_part = MIMEApplication("", _encoder=encode_7or8bit)
-        # TODO: encrypt() can return empty if a recipient's key is not signed. We should handle this
-        #   with a proper exception.
-        encrypted_part.set_payload(str(self.gpg.encrypt(signed_message.as_string(), fingerprint_list)))
-
-        # Pack it all into one big message
-        encrypted_message = MIMEMultipart(_subtype="encrypted", protocol="application/pgp-encrypted")
-        encrypted_message['Subject'] = message_dict['subject']
-        encrypted_message.attach(pgp_version)
-        encrypted_message.attach(encrypted_part)
-
-        return encrypted_message
-
-    def sendmail(self, message_dict):
-        # Use our magic
-        encrypted_message_string = str(self._eldtritch_crypto_magic(message_dict))
-
-        # Get a list of recipients from config
-        recipients = []
-        for recipient in self.config['recipients']:
-            recipients.append(recipient['email'])
-            self.logger.trace(recipient['email'])
-
-        # Mail servers will probably deauth you after a fixed period of inactivity.
-        # TODO: There is probably also a hard session limit too.
-        if (time.time() - self.lastSentTime) > self.config['smtp_max_idle']:
-            self.logger.info("Assuming the connection is dead.")
-            self._connect()
+        self.gpgkeyring = gpgkeyring
+        self.gpgkeyverifier = gpgkeyverifier
+        self.gpgmailbuilder = gpgmailbuilder.GpgMailBuilder(self.gpgkeyring, self.config['main_loop_duration'])
+        self.mailsender = mailsender.MailSender(self.config)
+
+        self.outbox_path = os.path.join(self.config['watch_dir'], 'outbox')
+
+        # Set expiration_warning_message so that it is not detected as changed immediately
+        #   after gpgmailer-daemon sends an email.
+        self.expiration_warning_message = gpgkeyverifier.get_expiration_warning_message(time.time())
 
+        self.logger.info('Done initializing gpgmailer module.')
+
+    # Gpgmailer's main loop. Reads the watch directory and then calls other modules
+    #   to build and send email.
+    def start_monitoring(self):
         try:
-            self.smtp.sendmail(self.config['sender']['email'], recipients, encrypted_message_string)
+            while True:
+                loop_start_time = time.time()
+
+                self.recipients = self.gpgkeyverifier.get_valid_recipients(loop_start_time)
+                self.keys = self.gpgkeyverifier.get_valid_keys(loop_start_time)
+
+                new_expiration_warning_message = self.gpgkeyverifier.get_expiration_warning_message(loop_start_time)
+                self._update_expiration_warning_message(new_expiration_warning_message)
+                self._send_email_if_new_expiration_warning_messages()
```

Comment:
They could. I thought that it would be easier to have two separate methods than try to sensibly name one that does two very different things, and there really doesn't seem to be much to be gained by combining them.